ICE: 'index out of bounds: the len is 8 but the index is 8', compiler/rustc_infer/src/infer/error_reporting/nice_region_error/util.rs:61:22

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.

http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/

-->

### Code

```Rust
fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.54.0-nightly (dbe459ded 2021-06-02)
binary: rustc
commit-hash: dbe459dedd33470f2cb28101157de316caaffa66
commit-date: 2021-06-02
host: x86_64-unknown-linux-gnu
release: 1.54.0-nightly
LLVM version: 12.0.1
```

### Error output

```
error: expected type, found `,`
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:47
  |
1 | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : FnO...
  |                                               ^ expected type

error: unexpected `self` parameter in function
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:51
  |
1 | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : FnO...
  |                                                   ^^^^ must be the first parameter of an associated function

error: unexpected `self` parameter in function
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:58
  |
1 | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : FnO...
  |                                                          ^^^^ must be the first parameter of an associated function

error: unexpected `self` parameter in function
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:67
  |
1 | ...            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
  |                                    ^^^^ must be the first parameter of an associated function

error: unexpected `self` parameter in function
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:76
  |
1 | ...   ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
  |                                    ^^^^ must be the first parameter of an associated function

error: unexpected `self` parameter in function
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:91
  |
1 | ...elf ,   self ,   self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
  |                                    ^^^^ must be the first parameter of an associated function

error: unexpected `self` parameter in function
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:100
  |
1 | ...elf ,   self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
  |                                    ^^^^ must be the first parameter of an associated function

error: `...` must be the last argument of a C-variadic function
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:83
  |
1 | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : FnO...
  |                                                                                   ^^^^

error: only foreign or `unsafe extern "C" functions may be C-variadic
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:83
  |
1 | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : FnO...
  |                                                                                   ^^^^

error: only foreign or `unsafe extern "C" functions may be C-variadic
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:107
  |
1 | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : FnO...
  |                                                                                                           ^^^^

error[E0412]: cannot find type `F` in this scope
  --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:119
   |
1  |   fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : ...
   |                                                                                                                         ^
   |
help: a trait with a similar name exists
   |
1  | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where Fn : FnOnce ( & 'a & 'b usize ) {
   |                                                                                                                       ^^
help: you might be missing a type parameter
   |
1  | fn ordering4 < 'a , 'b, F     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
   |                       ^^^

error[E0658]: C-variadic functions are unstable
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:1
  |
1 | / fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : F...
2 | | }
  | |_^
  |
  = note: see issue #44930 <https://github.com/rust-lang/rust/issues/44930> for more information
  = help: add `#![feature(c_variadic)]` to the crate attributes to enable

thread 'rustc' panicked at 'index out of bounds: the len is 8 but the index is 8', compiler/rustc_infer/src/infer/error_reporting/nice_region_error/util.rs:61:22
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.54.0-nightly (dbe459ded 2021-06-02) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z mir-opt-level=3 -Z new-llvm-pass-manager=yes -Z unsound-mir-opts=yes -Z polonius=yes -Z polymorphize=yes -C debuginfo=2 -C opt-level=3 --crate-type staticlib

query stack during panic:
#0 [check_item_well_formed] checking that `ordering4` is well-formed
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 12 previous errors

Some errors have detailed explanations: E0412, E0658.
For more information about an error, try `rustc --explain E0412`.
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
error: expected type, found `,`
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:47
  |
1 | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : FnO...
  |                                               ^ expected type

error: unexpected `self` parameter in function
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:51
  |
1 | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : FnO...
  |                                                   ^^^^ must be the first parameter of an associated function

error: unexpected `self` parameter in function
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:58
  |
1 | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : FnO...
  |                                                          ^^^^ must be the first parameter of an associated function

error: unexpected `self` parameter in function
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:67
  |
1 | ...            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
  |                                    ^^^^ must be the first parameter of an associated function

error: unexpected `self` parameter in function
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:76
  |
1 | ...   ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
  |                                    ^^^^ must be the first parameter of an associated function

error: unexpected `self` parameter in function
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:91
  |
1 | ...elf ,   self ,   self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
  |                                    ^^^^ must be the first parameter of an associated function

error: unexpected `self` parameter in function
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:100
  |
1 | ...elf ,   self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
  |                                    ^^^^ must be the first parameter of an associated function

error: `...` must be the last argument of a C-variadic function
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:83
  |
1 | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : FnO...
  |                                                                                   ^^^^

error: only foreign or `unsafe extern "C" functions may be C-variadic
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:83
  |
1 | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : FnO...
  |                                                                                   ^^^^

error: only foreign or `unsafe extern "C" functions may be C-variadic
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:107
  |
1 | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : FnO...
  |                                                                                                           ^^^^

error[E0412]: cannot find type `F` in this scope
  --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:119
   |
1  |   fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : ...
   |                                                                                                                         ^
   |
help: a trait with a similar name exists
   |
1  | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where Fn : FnOnce ( & 'a & 'b usize ) {
   |                                                                                                                       ^^
help: you might be missing a type parameter
   |
1  | fn ordering4 < 'a , 'b, F     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
   |                       ^^^

error[E0658]: C-variadic functions are unstable
 --> default_finding_folder_rust/crash_20210604_133458_002a/delta/perses_result/mutant.rs:1:1
  |
1 | / fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,   self , ... ,   self ,   self , ... ) where F : F...
2 | | }
  | |_^
  |
  = note: see issue #44930 <https://github.com/rust-lang/rust/issues/44930> for more information
  = help: add `#![feature(c_variadic)]` to the crate attributes to enable

thread 'rustc' panicked at 'index out of bounds: the len is 8 but the index is 8', compiler/rustc_infer/src/infer/error_reporting/nice_region_error/util.rs:61:22
stack backtrace:
   0: rust_begin_unwind
             at /rustc/dbe459dedd33470f2cb28101157de316caaffa66/library/std/src/panicking.rs:515:5
   1: core::panicking::panic_fmt
             at /rustc/dbe459dedd33470f2cb28101157de316caaffa66/library/core/src/panicking.rs:92:14
   2: core::panicking::panic_bounds_check
             at /rustc/dbe459dedd33470f2cb28101157de316caaffa66/library/core/src/panicking.rs:69:5
   3: rustc_infer::infer::error_reporting::nice_region_error::util::<impl rustc_infer::infer::error_reporting::nice_region_error::NiceRegionError>::find_param_with_region
   4: rustc_infer::infer::error_reporting::nice_region_error::NiceRegionError::try_report_from_nll
   5: rustc_infer::infer::error_reporting::nice_region_error::NiceRegionError::try_report
   6: rustc_infer::infer::error_reporting::nice_region_error::<impl rustc_infer::infer::InferCtxt>::try_report_nice_region_error
   7: rustc_infer::infer::error_reporting::<impl rustc_infer::infer::InferCtxt>::report_region_errors
   8: rustc_infer::infer::InferCtxt::resolve_regions_and_report_errors
   9: rustc_typeck::check::regionck::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::regionck_item
  10: rustc_typeck::check::inherited::InheritedBuilder::enter
  11: rustc_typeck::check::wfcheck::check_item_fn
  12: rustc_typeck::check::wfcheck::check_item_well_formed
  13: rustc_query_system::query::plumbing::get_query_impl
  14: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::check_item_well_formed
  15: <rustc_typeck::check::wfcheck::CheckTypeWellFormedVisitor as rustc_hir::intravisit::Visitor>::visit_item
  16: rustc_data_structures::sync::par_for_each_in
  17: rustc_hir::hir::Crate::par_visit_all_item_likes
  18: rustc_session::session::Session::track_errors
  19: rustc_typeck::check_crate
  20: rustc_interface::passes::analysis
  21: rustc_query_system::query::plumbing::get_query_impl
  22: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::analysis
  23: rustc_interface::passes::QueryContext::enter
  24: rustc_span::with_session_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.54.0-nightly (dbe459ded 2021-06-02) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z mir-opt-level=3 -Z new-llvm-pass-manager=yes -Z unsound-mir-opts=yes -Z polonius=yes -Z polymorphize=yes -C debuginfo=2 -C opt-level=3 --crate-type staticlib

query stack during panic:
#0 [check_item_well_formed] checking that `ordering4` is well-formed
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 12 previous errors

Some errors have detailed explanations: E0412, E0658.
For more information about an error, try `rustc --explain E0412`.
```

</p>
</details>

--- a/compiler/rustc_ast_passes/src/ast_validation.rs
+++ b/compiler/rustc_ast_passes/src/ast_validation.rs
@@ -652,7 +652,7 @@ impl<'a> AstValidator<'a> {
                 self.err_handler()
                     .struct_span_err(
                         *span,
-                        "only foreign or `unsafe extern \"C\" functions may be C-variadic",
+                        "only foreign or `unsafe extern \"C\"` functions may be C-variadic",
                     )
                     .emit();
             }

--- a/compiler/rustc_infer/src/infer/error_reporting/nice_region_error/util.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/nice_region_error/util.rs
@@ -56,33 +56,42 @@ impl<'a, 'tcx> NiceRegionError<'a, 'tcx> {
         let fn_decl = hir.fn_decl_by_hir_id(owner_id).unwrap();
         let poly_fn_sig = self.tcx().fn_sig(id);
         let fn_sig = self.tcx().liberate_late_bound_regions(id, poly_fn_sig);
-        body.params.iter().enumerate().find_map(|(index, param)| {
-            // May return None; sometimes the tables are not yet populated.
-            let ty = fn_sig.inputs()[index];
-            let mut found_anon_region = false;
-            let new_param_ty = self.tcx().fold_regions(ty, &mut false, |r, _| {
-                if *r == *anon_region {
-                    found_anon_region = true;
-                    replace_region
-                } else {
-                    r
-                }
-            });
-            if found_anon_region {
-                let ty_hir_id = fn_decl.inputs[index].hir_id;
-                let param_ty_span = hir.span(ty_hir_id);
-                let is_first = index == 0;
-                Some(AnonymousParamInfo {
-                    param,
-                    param_ty: new_param_ty,
-                    param_ty_span,
-                    bound_region,
-                    is_first,
-                })
+        body.params
+            .iter()
+            .take(if fn_sig.c_variadic {
+                fn_sig.inputs().len()
             } else {
-                None
-            }
-        })
+                assert_eq!(fn_sig.inputs().len(), body.params.len());
+                body.params.len()
+            })
+            .enumerate()
+            .find_map(|(index, param)| {
+                // May return None; sometimes the tables are not yet populated.
+                let ty = fn_sig.inputs()[index];
+                let mut found_anon_region = false;
+                let new_param_ty = self.tcx().fold_regions(ty, &mut false, |r, _| {
+                    if *r == *anon_region {
+                        found_anon_region = true;
+                        replace_region
+                    } else {
+                        r
+                    }
+                });
+                if found_anon_region {
+                    let ty_hir_id = fn_decl.inputs[index].hir_id;
+                    let param_ty_span = hir.span(ty_hir_id);
+                    let is_first = index == 0;
+                    Some(AnonymousParamInfo {
+                        param,
+                        param_ty: new_param_ty,
+                        param_ty_span,
+                        bound_region,
+                        is_first,
+                    })
+                } else {
+                    None
+                }
+            })
     }
 
     pub(super) fn future_return_type(

--- a/src/test/ui/c-variadic/issue-86053-1.rs
+++ b/src/test/ui/c-variadic/issue-86053-1.rs
@@ -1,0 +1,12 @@
+// Regression test for the ICE described in issue #86053.
+// error-pattern:unexpected `self` parameter in function
+// error-pattern:`...` must be the last argument of a C-variadic function
+// error-pattern:cannot find type `F` in this scope
+// error-pattern:in type `&'a &'b usize`, reference has a longer lifetime than the data it references
+
+#![feature(c_variadic)]
+#![crate_type="lib"]
+
+fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,
+    self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
+}

--- a/src/test/ui/c-variadic/issue-86053-1.stderr
+++ b/src/test/ui/c-variadic/issue-86053-1.stderr
@@ -1,0 +1,101 @@
+error: expected type, found `,`
+  --> $DIR/issue-86053-1.rs:10:47
+   |
+LL | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,
+   |                                               ^ expected type
+
+error: unexpected `self` parameter in function
+  --> $DIR/issue-86053-1.rs:10:51
+   |
+LL | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,
+   |                                                   ^^^^ must be the first parameter of an associated function
+
+error: unexpected `self` parameter in function
+  --> $DIR/issue-86053-1.rs:10:58
+   |
+LL | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,
+   |                                                          ^^^^ must be the first parameter of an associated function
+
+error: unexpected `self` parameter in function
+  --> $DIR/issue-86053-1.rs:10:67
+   |
+LL | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,
+   |                                                                   ^^^^ must be the first parameter of an associated function
+
+error: unexpected `self` parameter in function
+  --> $DIR/issue-86053-1.rs:11:5
+   |
+LL |     self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
+   |     ^^^^ must be the first parameter of an associated function
+
+error: unexpected `self` parameter in function
+  --> $DIR/issue-86053-1.rs:11:20
+   |
+LL |     self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
+   |                    ^^^^ must be the first parameter of an associated function
+
+error: unexpected `self` parameter in function
+  --> $DIR/issue-86053-1.rs:11:29
+   |
+LL |     self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
+   |                             ^^^^ must be the first parameter of an associated function
+
+error: `...` must be the last argument of a C-variadic function
+  --> $DIR/issue-86053-1.rs:11:12
+   |
+LL |     self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
+   |            ^^^^
+
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
+  --> $DIR/issue-86053-1.rs:11:12
+   |
+LL |     self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
+   |            ^^^^
+
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
+  --> $DIR/issue-86053-1.rs:11:36
+   |
+LL |     self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
+   |                                    ^^^^
+
+error[E0412]: cannot find type `F` in this scope
+  --> $DIR/issue-86053-1.rs:11:48
+   |
+LL |     self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
+   |                                                ^
+   | 
+  ::: $SRC_DIR/core/src/ops/function.rs:LL:COL
+   |
+LL | pub trait Fn<Args>: FnMut<Args> {
+   | ------------------------------- similarly named trait `Fn` defined here
+   |
+help: a trait with a similar name exists
+   |
+LL |     self , ... ,   self ,   self , ... ) where Fn : FnOnce ( & 'a & 'b usize ) {
+   |                                                ^^
+help: you might be missing a type parameter
+   |
+LL | fn ordering4 < 'a , 'b, F     > ( a :            ,   self , self ,   self ,
+   |                       ^^^
+
+error[E0491]: in type `&'a &'b usize`, reference has a longer lifetime than the data it references
+  --> $DIR/issue-86053-1.rs:11:52
+   |
+LL |     self , ... ,   self ,   self , ... ) where F : FnOnce ( & 'a & 'b usize ) {
+   |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+note: the pointer is valid for the lifetime `'a` as defined on the function body at 10:16
+  --> $DIR/issue-86053-1.rs:10:16
+   |
+LL | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,
+   |                ^^
+note: but the referenced data is only valid for the lifetime `'b` as defined on the function body at 10:21
+  --> $DIR/issue-86053-1.rs:10:21
+   |
+LL | fn ordering4 < 'a , 'b     > ( a :            ,   self , self ,   self ,
+   |                     ^^
+
+error: aborting due to 12 previous errors
+
+Some errors have detailed explanations: E0412, E0491.
+For more information about an error, try `rustc --explain E0412`.

--- a/src/test/ui/c-variadic/issue-86053-2.rs
+++ b/src/test/ui/c-variadic/issue-86053-2.rs
@@ -1,0 +1,11 @@
+// Regression test for the ICE caused by the example in
+// https://github.com/rust-lang/rust/issues/86053#issuecomment-855672258
+
+#![feature(c_variadic)]
+
+trait H<T> {}
+
+unsafe extern "C" fn ordering4<'a, F: H<&'static &'a ()>>(_: (), ...) {}
+//~^ ERROR: in type `&'static &'a ()`, reference has a longer lifetime than the data it references [E0491]
+
+fn main() {}

--- a/src/test/ui/c-variadic/issue-86053-2.stderr
+++ b/src/test/ui/c-variadic/issue-86053-2.stderr
@@ -1,0 +1,16 @@
+error[E0491]: in type `&'static &'a ()`, reference has a longer lifetime than the data it references
+  --> $DIR/issue-86053-2.rs:8:39
+   |
+LL | unsafe extern "C" fn ordering4<'a, F: H<&'static &'a ()>>(_: (), ...) {}
+   |                                       ^^^^^^^^^^^^^^^^^^
+   |
+   = note: the pointer is valid for the static lifetime
+note: but the referenced data is only valid for the lifetime `'a` as defined on the function body at 8:32
+  --> $DIR/issue-86053-2.rs:8:32
+   |
+LL | unsafe extern "C" fn ordering4<'a, F: H<&'static &'a ()>>(_: (), ...) {}
+   |                                ^^
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0491`.

--- a/src/test/ui/mir/issue-83499-input-output-iteration-ice.rs
+++ b/src/test/ui/mir/issue-83499-input-output-iteration-ice.rs
@@ -5,6 +5,6 @@
 fn main() {}
 
 fn foo(_: Bar, ...) -> impl {}
-//~^ ERROR only foreign or `unsafe extern "C" functions may be C-variadic
+//~^ ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
 //~| ERROR cannot find type `Bar` in this scope
 //~| ERROR at least one trait must be specified

--- a/src/test/ui/mir/issue-83499-input-output-iteration-ice.stderr
+++ b/src/test/ui/mir/issue-83499-input-output-iteration-ice.stderr
@@ -1,4 +1,4 @@
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/issue-83499-input-output-iteration-ice.rs:7:16
    |
 LL | fn foo(_: Bar, ...) -> impl {}

--- a/src/test/ui/parser/variadic-ffi-semantic-restrictions.rs
+++ b/src/test/ui/parser/variadic-ffi-semantic-restrictions.rs
@@ -4,32 +4,32 @@
 fn main() {}
 
 fn f1_1(x: isize, ...) {}
-//~^ ERROR only foreign or `unsafe extern "C" functions may be C-variadic
+//~^ ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
 
 fn f1_2(...) {}
-//~^ ERROR only foreign or `unsafe extern "C" functions may be C-variadic
+//~^ ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
 //~| ERROR C-variadic function must be declared with at least one named argument
 
 extern "C" fn f2_1(x: isize, ...) {}
-//~^ ERROR only foreign or `unsafe extern "C" functions may be C-variadic
+//~^ ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
 
 extern "C" fn f2_2(...) {}
-//~^ ERROR only foreign or `unsafe extern "C" functions may be C-variadic
+//~^ ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
 //~| ERROR C-variadic function must be declared with at least one named argument
 
 extern "C" fn f2_3(..., x: isize) {}
-//~^ ERROR only foreign or `unsafe extern "C" functions may be C-variadic
+//~^ ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
 //~| ERROR `...` must be the last argument of a C-variadic function
 
 extern "C" fn f3_1(x: isize, ...) {}
-//~^ ERROR only foreign or `unsafe extern "C" functions may be C-variadic
+//~^ ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
 
 extern "C" fn f3_2(...) {}
-//~^ ERROR only foreign or `unsafe extern "C" functions may be C-variadic
+//~^ ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
 //~| ERROR C-variadic function must be declared with at least one named argument
 
 extern "C" fn f3_3(..., x: isize) {}
-//~^ ERROR only foreign or `unsafe extern "C" functions may be C-variadic
+//~^ ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
 //~| ERROR `...` must be the last argument of a C-variadic function
 
 extern "C" {
@@ -43,35 +43,35 @@ struct X;
 
 impl X {
     fn i_f1(x: isize, ...) {}
-    //~^ ERROR only foreign or `unsafe extern "C" functions may be C-variadic
+    //~^ ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
     fn i_f2(...) {}
-    //~^ ERROR only foreign or `unsafe extern "C" functions may be C-variadic
+    //~^ ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
     //~| ERROR C-variadic function must be declared with at least one named argument
     fn i_f3(..., x: isize, ...) {}
-    //~^ ERROR only foreign or `unsafe extern "C" functions may be C-variadic
-    //~| ERROR only foreign or `unsafe extern "C" functions may be C-variadic
+    //~^ ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
+    //~| ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
     //~| ERROR `...` must be the last argument of a C-variadic function
     fn i_f4(..., x: isize, ...) {}
-    //~^ ERROR only foreign or `unsafe extern "C" functions may be C-variadic
-    //~| ERROR only foreign or `unsafe extern "C" functions may be C-variadic
+    //~^ ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
+    //~| ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
     //~| ERROR `...` must be the last argument of a C-variadic function
 }
 
 trait T {
     fn t_f1(x: isize, ...) {}
-    //~^ ERROR only foreign or `unsafe extern "C" functions may be C-variadic
+    //~^ ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
     fn t_f2(x: isize, ...);
-    //~^ ERROR only foreign or `unsafe extern "C" functions may be C-variadic
+    //~^ ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
     fn t_f3(...) {}
-    //~^ ERROR only foreign or `unsafe extern "C" functions may be C-variadic
+    //~^ ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
     //~| ERROR C-variadic function must be declared with at least one named argument
     fn t_f4(...);
-    //~^ ERROR only foreign or `unsafe extern "C" functions may be C-variadic
+    //~^ ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
     //~| ERROR C-variadic function must be declared with at least one named argument
     fn t_f5(..., x: isize) {}
-    //~^ ERROR only foreign or `unsafe extern "C" functions may be C-variadic
+    //~^ ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
     //~| ERROR `...` must be the last argument of a C-variadic function
     fn t_f6(..., x: isize);
-    //~^ ERROR only foreign or `unsafe extern "C" functions may be C-variadic
+    //~^ ERROR only foreign or `unsafe extern "C"` functions may be C-variadic
     //~| ERROR `...` must be the last argument of a C-variadic function
 }

--- a/src/test/ui/parser/variadic-ffi-semantic-restrictions.stderr
+++ b/src/test/ui/parser/variadic-ffi-semantic-restrictions.stderr
@@ -1,4 +1,4 @@
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:6:19
    |
 LL | fn f1_1(x: isize, ...) {}
@@ -10,13 +10,13 @@ error: C-variadic function must be declared with at least one named argument
 LL | fn f1_2(...) {}
    |         ^^^
 
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:9:9
    |
 LL | fn f1_2(...) {}
    |         ^^^
 
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:13:30
    |
 LL | extern "C" fn f2_1(x: isize, ...) {}
@@ -28,7 +28,7 @@ error: C-variadic function must be declared with at least one named argument
 LL | extern "C" fn f2_2(...) {}
    |                    ^^^
 
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:16:20
    |
 LL | extern "C" fn f2_2(...) {}
@@ -40,13 +40,13 @@ error: `...` must be the last argument of a C-variadic function
 LL | extern "C" fn f2_3(..., x: isize) {}
    |                    ^^^
 
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:20:20
    |
 LL | extern "C" fn f2_3(..., x: isize) {}
    |                    ^^^
 
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:24:30
    |
 LL | extern "C" fn f3_1(x: isize, ...) {}
@@ -58,7 +58,7 @@ error: C-variadic function must be declared with at least one named argument
 LL | extern "C" fn f3_2(...) {}
    |                    ^^^
 
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:27:20
    |
 LL | extern "C" fn f3_2(...) {}
@@ -70,7 +70,7 @@ error: `...` must be the last argument of a C-variadic function
 LL | extern "C" fn f3_3(..., x: isize) {}
    |                    ^^^
 
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:31:20
    |
 LL | extern "C" fn f3_3(..., x: isize) {}
@@ -88,7 +88,7 @@ error: `...` must be the last argument of a C-variadic function
 LL |     fn e_f2(..., x: isize);
    |             ^^^
 
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:45:23
    |
 LL |     fn i_f1(x: isize, ...) {}
@@ -100,7 +100,7 @@ error: C-variadic function must be declared with at least one named argument
 LL |     fn i_f2(...) {}
    |             ^^^
 
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:47:13
    |
 LL |     fn i_f2(...) {}
@@ -112,13 +112,13 @@ error: `...` must be the last argument of a C-variadic function
 LL |     fn i_f3(..., x: isize, ...) {}
    |             ^^^
 
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:50:13
    |
 LL |     fn i_f3(..., x: isize, ...) {}
    |             ^^^
 
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:50:28
    |
 LL |     fn i_f3(..., x: isize, ...) {}
@@ -130,25 +130,25 @@ error: `...` must be the last argument of a C-variadic function
 LL |     fn i_f4(..., x: isize, ...) {}
    |             ^^^
 
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:54:13
    |
 LL |     fn i_f4(..., x: isize, ...) {}
    |             ^^^
 
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:54:28
    |
 LL |     fn i_f4(..., x: isize, ...) {}
    |                            ^^^
 
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:61:23
    |
 LL |     fn t_f1(x: isize, ...) {}
    |                       ^^^
 
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:63:23
    |
 LL |     fn t_f2(x: isize, ...);
@@ -160,7 +160,7 @@ error: C-variadic function must be declared with at least one named argument
 LL |     fn t_f3(...) {}
    |             ^^^
 
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:65:13
    |
 LL |     fn t_f3(...) {}
@@ -172,7 +172,7 @@ error: C-variadic function must be declared with at least one named argument
 LL |     fn t_f4(...);
    |             ^^^
 
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:68:13
    |
 LL |     fn t_f4(...);
@@ -184,7 +184,7 @@ error: `...` must be the last argument of a C-variadic function
 LL |     fn t_f5(..., x: isize) {}
    |             ^^^
 
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:71:13
    |
 LL |     fn t_f5(..., x: isize) {}
@@ -196,7 +196,7 @@ error: `...` must be the last argument of a C-variadic function
 LL |     fn t_f6(..., x: isize);
    |             ^^^
 
-error: only foreign or `unsafe extern "C" functions may be C-variadic
+error: only foreign or `unsafe extern "C"` functions may be C-variadic
   --> $DIR/variadic-ffi-semantic-restrictions.rs:74:13
    |
 LL |     fn t_f6(..., x: isize);

